src/persist/src/operators/stream.rs

```diff
@@ -466,4 +660,198 @@ mod tests {
+        };
+        {
+            let listen_fn = Box::new(move |e| {
+                match e {
+                    ListenEvent::Sealed(t) => listen_tx.send(Sealed::Condition(t)).unwrap(),
+                    _ => panic!("unexpected data"),
+                };
+                ()
+            });
+
+            condition_read.listen(listen_fn)?;
+        };
+
+        timely::execute_directly(move |worker| {
+            let (mut primary_input, mut condition_input, primary_probe, condition_probe) = worker
+                .dataflow(|scope| {
+                    let (primary_write, _read) = p.create_or_load::<(), ()>("primary").unwrap();
+                    let (condition_write, _read) = p.create_or_load::<(), ()>("condition").unwrap();
+                    let mut primary_input = Handle::new();
+                    let mut condition_input = Handle::new();
+                    let primary_stream = primary_input.to_stream(scope);
+                    let condition_stream = condition_input.to_stream(scope);
+                    let (_, _) = primary_stream.conditional_seal(
+                        "test",
+                        &condition_stream,
+                        primary_write,
+                        condition_write,
+                    );
+
+                    let primary_probe = primary_stream.probe();
+                    let condition_probe = condition_stream.probe();
+
+                    (
+                        primary_input,
+                        condition_input,
+                        primary_probe,
+                        condition_probe,
+                    )
+                });
+
+            // Only send data on the condition input, not on the primary input. This simulates the
+            // case where our primary input never sees any data.
+            condition_input.send((((), ()), 0, 1));
+
+            primary_input.advance_to(1);
+            condition_input.advance_to(1);
+            while primary_probe.less_than(&1) {
+                worker.step();
+            }
+
+            primary_input.advance_to(3);
+            while primary_probe.less_than(&3) {
+                worker.step();
+            }
+
+            condition_input.advance_to(2);
+            while condition_probe.less_than(&2) {
+                worker.step();
+            }
+
+            condition_input.advance_to(3);
+            while condition_probe.less_than(&3) {
+                worker.step();
+            }
+        });
+
+        let actual_seals: Vec<_> = listen_rx.try_iter().collect();
+
+        // Assert that:
+        //  a) We don't seal primary when condition has not sufficiently advanced.
+        //  b) Condition is sealed before primary for the same timestamp.
+        //  c) We seal up, even when never receiving any data.
+        assert_eq!(
+            vec![
+                Sealed::Condition(1),
+                Sealed::Primary(1),
+                Sealed::Condition(2),
```

Comment:
Okay so walk me though why there isn't a `Sealed::Primary(2)` between `Sealed::Condition(2)` and `Sealed::Condition(3)`. I don't immediately see how this works

Reply:
This has to do with how I'm advancing input frontiers in the test. I'm advancing the frontier on primary to 3, skipping 2. When the operator does a seal, it looks at what the current input frontier is and sets that as the next notify. In practice, the frontier for both inputs will advance in lock-step because they both originate from the same source operator in the end (in our persistent source use case).

@ruchirK This is also a partial explanation for your question below, about `frontier(0)` and `frontier(1)`. When setting the next notify, I'm looking at the input frontier for each input separately. I could sync them up internally by looking at the combined frontier when setting the next notify. This would also mean that a `Sealed::Primary(2)` would show up in the test.